api/update.go

```diff
@@ -40,73 +25,73 @@ func (api API) Update(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, msg, http.StatusInternalServerError)
 		return
 	}
-	var actualPreload []preloadlist.Entry
-	for _, entry := range preloadList.Entries {
-		if entry.Mode == preloadlist.ForceHTTPS {
-			actualPreload = append(actualPreload, entry)
-		}
-	}
 
-	// Get domains currently recorded as preloaded.
-	preloadedDomains, dbErr := api.database.StatesWithStatus(database.StatusPreloaded)
-	if dbErr != nil {
-		msg := fmt.Sprintf(
-			"Internal error: could not retrieve domain names previously marked as preloaded. (%s)\n",
-			dbErr,
-		)
-		http.Error(w, msg, http.StatusInternalServerError)
-		return
-	}
-	var databasePreload []preloadlist.Entry
-	for _, ds := range preloadedDomains {
-		databasePreload = append(databasePreload, ds.ToEntry())
+	domainStates := make(map[string]database.DomainState)
+	addDomainStatesWithStatus := func(status database.PreloadStatus) bool {
+		domains, err := api.database.StatesWithStatus(status)
+		if err != nil {
+			msg := fmt.Sprintf("Internal error: could not retrieve domain names previously marked as %s. (%s)\n", status, err)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return false
+		}
+		for _, domain := range domains {
+			domainStates[domain.Name] = domain
+		}
+		return true
 	}
-
-	// Get domains currently recorded as pending removal.
-	pendingRemovalDomains, dbErr := api.database.StatesWithStatus(database.StatusPendingRemoval)
-	if dbErr != nil {
-		msg := fmt.Sprintf(
-			"Internal error: could not retrieve domain names previously marked as pending removal. (%s)\n",
-			dbErr,
-		)
-		http.Error(w, msg, http.StatusInternalServerError)
+	// Get domains currently recorded as preloaded, pending removal, or
+	// pending automated removal.
+	if !addDomainStatesWithStatus(database.StatusPreloaded) ||
+		!addDomainStatesWithStatus(database.StatusPendingRemoval) ||
+		!addDomainStatesWithStatus(database.StatusPendingAutomatedRemoval) {
 		return
 	}
-	var databasePendingRemoval []preloadlist.Entry
-	for _, ds := range pendingRemovalDomains {
-		databasePendingRemoval = append(databasePendingRemoval, ds.ToEntry())
-	}
 
-	// Calculate values that are out of date.
 	var updates []database.DomainState
-
-	added := difference(difference(actualPreload, databasePreload), databasePendingRemoval)
-	for _, entry := range added {
-		updates = append(updates, database.EntryToDomainState(entry, database.StatusPreloaded))
-	}
-
-	removed := difference(databasePreload, actualPreload)
-	for _, entry := range removed {
-		updates = append(updates, database.EntryToDomainState(entry, database.StatusRemoved))
+	added := 0
+	updated := 0
+	removed := 0
+	for _, entry := range preloadList.Entries {
+		if entry.Mode != preloadlist.ForceHTTPS {
+			continue
+		}
+		domainState, found := domainStates[entry.Name]
+		if !found {
+			// entry is on the preload list but not marked as
+			// preloaded or pending removal in the database. Mark it
```

Comment:
Nit: preloaded, pending removal, or pending automated removal (for completeness sake)

Reply:
Done.